File: core/03-structured-output/go/main.go
// main.go —— demo only：三个场景（简历 / 产品 / 情感分类）。
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/joho/godotenv"
	"github.com/sashabaranov/go-openai"
)

func main() {
	if err := godotenv.Load("../.env"); err != nil {
		log.Println("warning: .env not loaded:", err)
	}

	cfg := openai.DefaultConfig(envOr("API_KEY", "not-needed"))
	cfg.BaseURL = os.Getenv("API_BASE_URL")
	cfg.HTTPClient = &http.Client{
		Transport: &http.Transport{Proxy: nil},
		Timeout:   60 * time.Second,
	}
	c := openai.NewClientWithConfig(cfg)
	model := os.Getenv("MODEL_ID")
	ctx := context.Background()

	run(ctx, c, model, "简历提取",
		"提取简历信息。",
		"张三，28岁，Python 工程师，邮箱 zs@example.com，擅长 Django、FastAPI、PostgreSQL。",
		"resume", schemaResume())

	run(ctx, c, model, "产品信息提取",
		"提取产品信息。",
		"iPhone 15 Pro 国行 9999 元，苹果出品，目前有货。",
		"product", schemaProduct())

	run(ctx, c, model, "情感分类（label 限定 positive/neutral/negative）",
		"对文本做情感分类，给出 label / confidence(0-1) / 一句话 reason。",
		"这部电影完全是浪费时间，特效粗糙，剧情拖沓。",
		"sentiment", schemaSentiment())
}

func run(ctx context.Context, c *openai.Client, model, label, system, user, name string, schema map[string]any) {
	var out map[string]any
	if err := Extract(ctx, c, model, system, user, name, schema, &out); err != nil {
		fmt.Printf("\n>>> %s\n  ERROR: %v\n", label, err)
		return
	}
	pretty, _ := json.MarshalIndent(out, "", "  ")
	fmt.Printf("\n>>> %s\n%s\n", label, string(pretty))
}

func envOr(k, fallback string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return fallback
}

File: core/03-structured-output/rust/src/main.rs
//! main.rs —— demo only：三个场景（简历 / 产品 / 情感分类）。

mod client;
mod schemas;

use client::{extract, Config};
use std::env;

fn main() {
    dotenv::from_filename("../.env").ok();
    let cfg = Config {
        base_url: env::var("API_BASE_URL").expect("API_BASE_URL not set"),
        api_key: env::var("API_KEY").unwrap_or_else(|_| "not-needed".into()),
        model: env::var("MODEL_ID").expect("MODEL_ID not set"),
    };

    run(&cfg, "简历提取",
        "提取简历信息。",
        "张三，28岁，Python 工程师，邮箱 zs@example.com，擅长 Django、FastAPI、PostgreSQL。",
        "resume", &schemas::resume());

    run(&cfg, "产品信息提取",
        "提取产品信息。",
        "iPhone 15 Pro 国行 9999 元，苹果出品，目前有货。",
        "product", &schemas::product());

    run(&cfg, "情感分类（label 限定 positive/neutral/negative）",
        "对文本做情感分类，给出 label / confidence(0-1) / 一句话 reason。",
        "这部电影完全是浪费时间，特效粗糙，剧情拖沓。",
        "sentiment", &schemas::sentiment());
}

fn run(cfg: &Config, label: &str, system: &str, user: &str, name: &str, schema: &serde_json::Value) {
    match extract(cfg, system, user, name, schema) {
        Ok(v) => println!("\n>>> {label}\n{}", serde_json::to_string_pretty(&v).unwrap()),
        Err(e) => println!("\n>>> {label}\n  ERROR: {e}"),
    }
}
